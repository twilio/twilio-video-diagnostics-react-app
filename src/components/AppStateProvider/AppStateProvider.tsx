import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
