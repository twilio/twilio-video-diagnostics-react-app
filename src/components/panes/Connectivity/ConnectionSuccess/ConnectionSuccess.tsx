import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { createStyles, makeStyles, Button, Container, Grid, Typography } from '@material-ui/core';
import Success from './Success.png';
import { ViewIcon } from '../../../../icons/ViewIcon';

const useStyles = makeStyles((theme) =>
  createStyles({
    illustrationContainer: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
    },
    viewButton: {
      marginTop: '2em',
      '& svg': {
        position: 'relative',
        left: '-5px',
      },
    },
  })
);

interface ConnectionSuccessProps {
  openModal: () => void;
}

export function ConnectionSuccess({ openModal }: ConnectionSuccessProps) {
  const classes = useStyles();
  const { nextPane } = useAppStateContext();

  return (
    <>
      <Container>
        <Grid container alignItems="center" justifyContent="space-between">
          <Grid item md={5}>
            <Typography variant="h1" gutterBottom>
              Connection success
            </Typography>

            <Typography variant="body1" gutterBottom>
              All connections to Twilio's servers are working correctly.
            </Typography>

            <Button variant="contained" color="primary" onClick={nextPane}>
              Ok
            </Button>
          </Grid>

          <Grid item md={5} className={classes.illustrationContainer}>
            {/* 
          The size of the image is explicitly stated here so that this content can properly be centered vertically
          before the image is loaded.
          */}
            <img src={Success} alt="Success" style={{ width: '245px', height: '200px' }} />
            <Button variant="outlined" onClick={openModal} className={classes.viewButton}>
              <ViewIcon />
              View detailed connection information
            </Button>
          </Grid>
        </Grid>
      </Container>
    </>
  );
}
