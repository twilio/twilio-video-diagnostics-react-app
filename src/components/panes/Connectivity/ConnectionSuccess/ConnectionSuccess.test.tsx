import { shallow } from 'enzyme';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { ConnectionSuccess } from './ConnectionSuccess';

jest.mock('../../../AppStateProvider/AppStateProvider');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;

mockUseAppStateContext.mockImplementation(() => ({ state: { activePane: 4 }, nextPane: jest.fn() }));

const mockSetIsModalOpen = jest.fn();

const wrapper = shallow(<ConnectionSuccess openModal={mockSetIsModalOpen} />);

describe('the ConnectionSuccess component', () => {
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
