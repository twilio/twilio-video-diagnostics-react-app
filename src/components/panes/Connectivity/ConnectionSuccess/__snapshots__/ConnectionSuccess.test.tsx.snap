// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ConnectionSuccess component should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Container))>
    <div>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-header-1"
        item={true}
        lg={5}
      >
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          variant="h1"
        >
          Connection success
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          variant="body1"
        >
          All connections to Twilio's servers are working correctly.
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-illustrationContainer-2"
        item={true}
        lg={5}
      >
        <img
          alt="Success"
          className="makeStyles-illustration-3"
          src="test-file-stub"
        />
        <WithStyles(ForwardRef(Button))
          className="makeStyles-viewButton-4"
          onClick={[MockFunction]}
          variant="outlined"
        >
          <ViewIcon />
          View detailed connection information
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-okButton-5"
        item={true}
        lg={5}
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          onClick={[MockFunction]}
          variant="contained"
        >
          Ok
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
    </div>
  </WithStyles(ForwardRef(Container))>
</Fragment>
`;
